python/tskit/combinatorics.py

```diff
@@ -0,0 +1,352 @@
+        r_labels = [lab for lab in labels if lab not in set(l_labels)]
+
+        left_id = _unrank(l_shape_rank, l_label_rank, l_labels)
+        right_id = _unrank(r_shape_rank, r_label_rank, r_labels)
+
+        # Arbitrarily setting time to be the number of leaves under the node
+        parent_id = tables.nodes.add_row(time=n)
+        tables.edges.add_row(0, seq_length, parent_id, left_id)
+        tables.edges.add_row(0, seq_length, parent_id, right_id)
+
+        return parent_id
+
+    for _ in range(n_total_leaves):
+        tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=0)
+
+    _unrank(shape_rank0, label_rank0, list(range(n_total_leaves)))
+
+    # Have to sort edges because we're doing a recursive, depth-first
+    # construction of this tree.
+    # To properly insert edges, we should be doing so breadth-first.
+    # However, we would prefer to keep this as simply recursive as
+    # possible until it's all tidy.
+    # So for now, we sweep our mistakes under the rug and sort.
+    tables.sort()
+    return tables.tree_sequence().first()
+
+
+def shape_unrank(rank, n):
+    return unrank(rank, 0, n)
+
+
+def label_unrank(tree, label_rank):
+    seq_length = tree.tree_sequence.sequence_length
+    tables = tskit.TableCollection(seq_length)
+
+    def _label_unrank(u, rank, labels, time):
+        """
+        Given the rank of a labeling, label this tree with the appropriate
+        labeling.
+        """
+        if tree.is_leaf(u):
+            assert len(labels) == 1
+            return labels[0]
+
+        left, right = canon_left_right(tree, u)
+
+        left_L = L_tree(tree, left)
+        right_L = L_tree(tree, right)
+
+        comb_rank = rank // (left_L * right_L)
+        rank_given_comb = rank % (left_L * right_L)
+
+        left_label_rank = rank_given_comb // right_L
+        right_label_rank = rank_given_comb % right_L
+
+        left_labels = Labeling.unrank(comb_rank, labels, tree.num_samples(left))
+        right_labels = [lab for lab in labels if lab not in set(left_labels)]
+
+        left_id = _label_unrank(left, left_label_rank, left_labels, time - 1)
+        right_id = _label_unrank(right, right_label_rank, right_labels, time - 1)
+
+        parent_id = tables.nodes.add_row(time=time)
+        tables.edges.add_row(0, seq_length, parent_id, left_id)
+        tables.edges.add_row(0, seq_length, parent_id, right_id)
+
+        return parent_id
+
+    n = tree.num_samples()
+    for _ in range(n):
+        tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=0)
+    _label_unrank(tree.root, label_rank, list(range(n)), n)
+
+    tables.sort()
+    return tables.tree_sequence().first()
+
+
+def S(n):
+    """
+    The cardinality of the set of unlabeled binary trees with n leaves,
+    under the conditions of tree isomorphism universal to this module.
+    """
+    if n <= 1:
+        return n
+
+    total = 0
+    for k in range(math.ceil(n / 2), n):
+        if k == n // 2:
```

Comment:
Seems inelegant to check this each time through the loop when it can only happen once, right? Is what you're doing something like
```python
total = 0
if n % 2 == 1:
    k = n // 2 + 1 
    total = (S(k) * (S(k) + 1)) // 2
for k in range(n // 2, n):
    total +=   S(k) * S(n - k)
```